### PR TITLE
Add Zkverse application

### PR DESCRIPTION
**Pull request by VegeBun-csj**
#### Project Abstract
At present, ZKP technology is constantly developing and innovating in the scaling and expansion track in Ethereum, especially some zk rollup projects, such as zksync, scroll, starknit, etc. So the main goal of this project is to introduce ZKP technology into the Polkadot/Substrate ecosystem. Polkadot/Substrate natively does not support ZKP, so this project(Zkverse, which means zk universe) will provide zk-related pallets to support substrate and more efficient zk proof generation ways.

There are three main goals for project: 
- Integrate some zkp libray(like bellman, plonk library and eg...) into substrate pallet
- Maximize the efficiency and convenience of zk proof generation(The proof is not generated on the chain, but through some developer friendly libraries(like snarkjs) which will be adapted to the zk lib on substrate-based chain.). This is very important and convenient for scaling. It can realize rollup for app-specific Dapps and greatly increase the throughput of the substrated-based chain. It can be said "Off-chain execution and on-chain verification".
- Publish some tutorial blogs/demos to let more substrate community developers enjoy the convenience of the above zkp development kits.

Meanwhile, We know that snakjs and circom are excellent and popular zkp development libraries. They are very popular in the Ethereum ecosystem and can automatically generate verification contract codes. We observed this very good feature, so we want to generate circuit codes more easily through circom, and generate proofs by snarkjs, which will be verified on substrate-based chain. Due to many people using snakjs, and circom is developer friendly, Unlike some domain-specific circuit writing methods, our project can attract many developers who are familiar with snarkjs to develop zkp Dapps in substrate/polkadot ecosystem. They can also enjoy the convenience of developing zkp applications in Polkadot ecosystem. So our main goal is to provide zkp infrastructure that is convenient for Polkadot developers. Also ,we will show a minimal example with Merkle tree and ZKP for rollup.

```
h(h(h(sm0+sm1) + h(sm2+sm3)) + sm4) (merkle root)
      /                         \
    h(h(sm0+sm1) + h(sm2+sm3))  sm4 (2 siblings)
    /          \                /
   h(sm0+sm1)   h(sm2+sm3)   sm4 (3 siblings)
  /   \        /   \         /
sm0   sm1   sm2    sm3      sm4 (leaves are the base level, 5 siblings)
^     ^     ^      ^        ^
|     |     |      |        |
m0    m1    m2     m3       m4
```

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by semuelle — CHANGES_REQUESTED** [review]
Hi @VegeBun-csj. Thank you for your application. I have a couple of follow-up questions:

- Are you planning to use these pallets yourself, or have you talked to anyone personally who would use them?
- You state that your project will "do more education" than the ZK-Snarks tutorial, but there are no education-related deliverables in your roadmap. If you meant the 0e deliverables (article/workshop), it would be great if you could expand their specification a bit. Given the price of the grant, I think it'd be justified to include detailed workshops from an introduction to zkp to implementation of example proofs.
- You already listed some similar projects, but did you check our [tech stack](https://wiki.polkadot.network/docs/build-open-source#chains-and-pallets) for other, existing implementations?

FYI, the Zeropool project is [still active](https://github.com/w3f/Grant-Milestone-Delivery/pull/316#issuecomment-1434023560).

**Comment by VegeBun-csj** [comment]
Thank you for your relay and guiding suggestions, and I will answer your questions separately.
> * Are you planning to use these pallets yourself, or have you talked to anyone personally who would use them?

Yes, I will also use this Pallet after developed. Also, we have talked with some developers before. They will also hope to develop some zkp dapps in Polkadot ecosystem. But at present, the threshold of using some zkp libraries is a little high, so we have made this project based on this pain point. I think we will promote it after we make it, and there should be many people interested. Because we will combine some relatively friendly libraries to facilitate developers.

> * You state that your project will "do more education" than the ZK-Snarks tutorial, but there are no education-related deliverables in your roadmap. If you meant the 0e deliverables (article/workshop), it would be great if you could expand their specification a bit. Given the price of the grant, I think it'd be justified to include detailed workshops from an introduction to zkp to implementation of example proofs.

Thank you for your suggestion. I have pushed the latest commit which involve the detailed plans of eucation. Also, Our example is not a simple circuit like `1+1`, but a relatively practical example, which can make developers feel more intuitively.

> * You already listed some similar projects, but did you check our [tech stack](https://wiki.polkadot.network/docs/build-open-source#chains-and-pallets) for other, existing implementations?
> FYI, the Zeropool project is [still active](https://github.com/w3f/Grant-Milestone-Delivery/pull/316#issuecomment-1434023560).

Yes, i have seen many similar projects, they do not implement our ideas. Our main idea is to make some libraries that are easy for developers to write circuit and generate zkp proof (such as snarkjs) compatible with some zkp libraries integrated on the substrate-based chain (these libraries need to develop relatively low-level circuits, which are not very friendly for developers to develop circuits and easy to make mistakes). At present, many projects use one library for proof generation and verification, so we did what they didn't do, which is developer friendly.

I just saw that they do have new actions, but their code base is relatively old. We intend to be compatible with the latest zkp library. In the future, we will also update Pallet to maintain the latest version compatibility with substrate and support developers.

File: applications/zkverse.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 / GPLv3 / MIT / Unlicense |
+| **0b.** | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an **article**/workshop to explain.
```
Comment (semuelle) [inline applications/zkverse.md:98]: Given that you have an "education" deliverable in each milestone, I'd say it makes sense to remove this article. Focus on the education part instead and, ideally, publish an article introducing the project separately for M2.

Comment (VegeBun-csj) [reply applications/zkverse.md]: > Given that you have an "education" deliverable in each milestone, I'd say it makes sense to remove this article. Focus on the education part instead and, ideally, publish an article introducing the project separately for M2.

thanks, i have modified it according to your suggestion

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the update, @VegeBun-csj. I am generally happy to proceed. I would just like to ask you to add the deliverables 0b (at least inline documentation) and 0c from our [template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) to both milestones, potentially also 0d (Docker) if the test setup is complex.

**Comment by VegeBun-csj** [comment]
> Thanks for the update, @VegeBun-csj. I am generally happy to proceed. I would just like to ask you to add the deliverables 0b (at least inline documentation) and 0c from our [template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) to both milestones, potentially also 0d (Docker) if the test setup is complex.

Thank you sir. I have updated:)

**Review by semuelle — DISMISSED** [review]
Thanks for the update.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks a lot for the application. Your first milestone says you will implement groth16, circom, and snarkjs. How is your solution different from https://github.com/bright/zk-snarks-with-substrate, which was updated three weeks ago and has the same components?

**Comment by VegeBun-csj** [comment]
> Thanks a lot for the application. Your first milestone says you will implement groth16, circom, and snarkjs. How is your solution different from https://github.com/bright/zk-snarks-with-substrate, which was updated three weeks ago and has the same components?

Thank you very much for your question. I saw [their implementation](https://github.com/bright/zk-snarks-with-substrate), but in fact, we had such an idea and implementation last year (earlier than them). You can check this [link](https://github.com/Zkvers/snarkjs-bellman-adapter). Secondly, compared with their implementation, we have some important advantages and differences:

1. Our implementation is to transform the proof generated by snarkjs (the transformed proof can be directly verified by the [bellman](https://github.com/zkcrypto/bellman) which will be compatiable with a verification pallet). The transformation of the snarkjs's proof is off-chain. This is a very important feature where the verification pallet on the chain can directly verify without other transformation. However, their implementation is to directly pass the proof of snarkjs to the chain which will be transformed on the chain firstly, and then verified, so it will occupy more on-chain resources.  We believe that the resources on the chain are very valuable, so we should do the more calculation off-chain as far as possible, and do the simplest calculation on the chain as far as possible.
2. Our implementation can be compatible with bellman and snarkjs at the same time. We have adapt the proof generated by the snarkjs but if anyone wants to use bellman to generate proof, they can also use our pallet to verify, so our implementation will be more general and their implementation will be limited.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the quick reply, and it’s nice to see that you are already working on this. However, we usually don’t fund work that is already been done, and previously we were unaware of your project, which is why we supported the other team. Sorry for this. Could you potentially reduce the price of the first milestone and/or focus more on the new features that you want to implement from now on?

**Comment by VegeBun-csj** [comment]
> Thanks for the quick reply, and it’s nice to see that you are already working on this. However, we usually don’t fund work that is already been done, and previously we were unaware of your project, which is why we supported the other team. Sorry for this. Could you potentially reduce the price of the first milestone and/or focus more on the new features that you want to implement from now on?

Thank you very much for your suggestion. Although I have completed some work before, the part of integrating pallet needs to be added and improved, eg: further writing pallet and optimize, version upgrade, function tests and some benchmarks. There is still a lot of work. At the same time, We will also add relevant educational content. It is not a simple example, such as [Simple operation](https://github.com/bright/zk-snarks-with-substrate/blob/main/circom/task.circom), but a more practical example (such as rollup or the other) to show devs how to use our tools to develop a more practical zk dapps which is more meaningful :)

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update and the price reduction. However, the current delivery for the first milestone could basically be a copy of the work done [here](https://github.com/bright/zk-snarks-with-substrate) based on the current specification. So I’m personally unwilling to accept it at this price point and given the current level of details, but others might have a different opinion.

**Comment by VegeBun-csj** [comment]
> Thanks for the update and the price reduction. However, the current delivery for the first milestone could basically be a copy of the work done [here](https://github.com/bright/zk-snarks-with-substrate) based on the current specification. So I’m personally unwilling to accept it at this price point and given the current level of details, but others might have a different opinion.

Thanks for your reply sincerely, but our implementation is not a copy. It is different from the details of the implementation. Can you reconsider my updated price reduction？

**Comment by VegeBun-csj** [comment]
Hi sir. I have reduced the price of milestone1 again. Can you take a look for my application?😃 
Thank you very much for your suggestions. @Noc2

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. But you changed the total costs of the first milestone to 70k USD (= 70,000). Was this intentional?

**Comment by VegeBun-csj** [comment]
> Thanks for the update. But you changed the total costs of the first milestone to 70k USD (= 70,000). Was this intentional?

I'm very sorry, sir. I didn't mean to. My total amount was reduced. It should have been caused by carelessness. I'm very sorry.

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I'm happy to go ahead with it.

**Review by keeganquigley — APPROVED** [review]
Thanks for the application @VegeBun-csj and for the deep dive @semuelle @Noc2 sounds like an interesting project. Hopefully these libraries will make zkp easier for devs. Happy to go forward with it also.

**Review by dsm-w3f — APPROVED** [review]
@VegeBun-csj I'm happy to support the project. The only part that I think you should be aware of is to focus more on the parts not covered for zk-SNARKs grant, i.e., bellman, plonk, etc, and when explaining the similar parts try to perform it with a different approach with different examples. I think this is a very useful project.

**Comment by keeganquigley** [comment]
Hi @VegeBun-csj how is milestone 2 coming along?

**Comment by keeganquigley** [comment]
pinging @VegeBun-csj

**Comment by keeganquigley** [comment]
Hi @VegeBun-csj if we don't hear back within 2 weeks we will close the grant due to inactivity. Let me know if you have any questions regarding this.

**Comment by VegeBun-csj** [comment]
@keeganquigley Hi, sir. I'm so sorry that i havent received any message from my side(like email or others). I missed it. I am still working on this project, but I encountered some obstacles before. May I apply to reopen this application? I think this is a very useful use case, and I hope I can complete it from start to finish, fulfill my wish 🙏

**Comment by keeganquigley** [comment]
Hi @VegeBun-csj thanks for your inquiry. Seeing that this grant is two years old and we canceled it a year ago, unfortunately you will need to apply for a new grant in order to reopen it. For budgeting purposes, we typically don't allow grants to run over multiple years. 

I'm sorry to bear this bad news, and wish you the best of luck with the project.

**Comment by VegeBun-csj** [comment]
@keeganquigley Thanks for your response.🙏

**Pull request merged**